### PR TITLE
[Feature] add BEiT

### PR DESCRIPTION
**Pull request by linfangjian01**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

add BEiT.

## Modification

add BEiT.

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Please add unit test in the couple of days.

It is better to add unit test together with support certain features in MMSegmentation.

**Comment by MengzhangLI** [comment]
Please fix the lint error.

File: mmseg/core/__init__.py [hunk 1]
```diff
@@ -1,4 +1,7 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from .evaluation import *  # noqa: F401, F403
+from .layer_decay_optimizer_constructor import \
+    LayerDecayOptimizerConstructor  # noqa: F401, F403
+from .resize_transform import SETR_Resize  # noqa: F401, F403
```
Comment (MeowZheng) [inline mmseg/core/__init__.py:5]: `# noqa: F401, F403` is ignore Flake8  F401 and F403. F401 is 'module imported but unused', and F403 is 'unable to detect undefined names '. I think F403 is useless here

File: mmseg/models/backbones/vit.py [hunk 2]
```diff
@@ -102,7 +102,6 @@ class VisionTransformer(BaseModule):
     This backbone is the implementation of `An Image is Worth 16x16 Words:
     Transformers for Image Recognition at
     Scale <https://arxiv.org/abs/2010.11929>`_.
-
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:105]: We may keep the new line

File: mmseg/core/layer_decay_optimizer_constructor.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return 0
+    elif var_name.startswith('backbone.layers'):
+        layer_id = int(var_name.split('.')[2])
+        return layer_id + 1
+    else:
+        return num_max_layer - 1
+
+
+@OPTIMIZER_BUILDERS.register_module()
+class LayerDecayOptimizerConstructor(DefaultOptimizerConstructor):
```
Comment (xvjiarui) [inline mmseg/core/layer_decay_optimizer_constructor.py:22]: Add description for this OptimizerConstructor.

File: mmseg/core/layer_decay_optimizer_constructor.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import json
+
+from mmcv.runner import (OPTIMIZER_BUILDERS, DefaultOptimizerConstructor,
+                         get_dist_info)
+
+
+def get_num_layer_for_vit(var_name, num_max_layer):
```
Comment (xvjiarui) [inline mmseg/core/layer_decay_optimizer_constructor.py:8]: We may add doc string for this function.

File: mmseg/core/layer_decay_optimizer_constructor.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            params (list[dict]): A list of param groups, it will be modified
+                in place.
+            module (nn.Module): The module to be added.
+            prefix (str): The prefix of the module
+            is_dcn_module (int|float|None): If the current module is a
+                submodule of DCN, `is_dcn_module` will be passed to
+                control conv_offset layer's learning rate. Defaults to None.
+        """
+        parameter_groups = {}
+        print(self.paramwise_cfg)
```
Comment (xvjiarui) [inline mmseg/core/layer_decay_optimizer_constructor.py:39]: use logger instead of print.

Comment (xvjiarui) [reply mmseg/core/layer_decay_optimizer_constructor.py:39]: `from mmseg.utils import get _root_logger`

File: mmseg/datasets/pipelines/resize_transform.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import mmcv
+import numpy as np
+
+from mmseg.datasets.builder import PIPELINES
+
+
+@PIPELINES.register_module()
+class SETR_Resize(object):
```
Comment (xvjiarui) [inline mmseg/datasets/pipelines/resize_transform.py:9]: Could we use the pipeline from SETR instead?

File: mmseg/models/backbones/beit.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        head_embed_dims = embed_dims // num_heads
+        self.scale = qk_scale or head_embed_dims**-0.5
+        if qkv_bias:
+            self.q_bias = nn.Parameter(torch.zeros(embed_dims))
+            self.v_bias = nn.Parameter(torch.zeros(embed_dims))
+        else:
+            self.q_bias = None
+            self.v_bias = None
```
Comment (xvjiarui) [inline mmseg/models/backbones/beit.py:62]: Should it be `qv_bias`?

File: mmseg/models/backbones/beit.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+                                                     src_size).float().numpy()
+                            f = interpolate.interp2d(x, y, z, kind='cubic')
+                            all_rel_pos_bias.append(
+                                torch.Tensor(f(dx, dy)).contiguous().view(
+                                    -1, 1).to(rel_pos_bias.device))
+
+                        rel_pos_bias = torch.cat(all_rel_pos_bias, dim=-1)
+                        new_rel_pos_bias = torch.cat(
+                            (rel_pos_bias, extra_tokens), dim=0)
+                        state_dict[key] = new_rel_pos_bias
```
Comment (MeowZheng) [inline mmseg/models/backbones/beit.py:397]: it is hard to understand, just add more comments for these lines

File: mmseg/models/necks/featurepyramid.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+@NECKS.register_module()
+class Feature2Pyramid(nn.Module):
+    """Feature2Pyramid.
+
+    A neck structure connect ViT backbone and decoder_heads.
+
+    Args:
+        embed_dims (int): embedding dimension.
+        rescales (list[float]): different sampling multiples were
+        used to obtain pyramid features. Default: (4, 2, 1, 0.5).
```
Comment (MeowZheng) [inline mmseg/models/necks/featurepyramid.py:16]: add indent

File: mmseg/models/backbones/beit.py [hunk 10]
```diff
@@ -0,0 +1,563 @@
...
+            self.q_bias = nn.Parameter(torch.zeros(embed_dims))
+            self.v_bias = nn.Parameter(torch.zeros(embed_dims))
+        else:
+            self.q_bias = None
+            self.v_bias = None
+
+        self.window_size = window_size
+        self.num_relative_distance = (2 * window_size[0] -
+                                        1) * (2 * window_size[1] - 1) + 3
+        # relative_position_bias_table shape is (2*Wh-1 * 2*Ww-1 + 3, nH)
```
Comment (MeowZheng) [inline mmseg/models/backbones/beit.py:67]: # 3 is for cls to token & token 2 cls & cls to cls

File: mmseg/models/backbones/beit.py [hunk 11]
```diff
@@ -0,0 +1,563 @@
...
+
+                if "relative_position_bias_table" in key:
+                    rel_pos_bias = state_dict[key]
+                    src_num_pos, num_attn_heads = rel_pos_bias.size()
+                    dst_num_pos, _ = self.state_dict()[key].size()
+                    dst_patch_shape = self.patch_shape
+                    if dst_patch_shape[0] != dst_patch_shape[1]:
+                        raise NotImplementedError()
+                    num_extra_tokens = dst_num_pos - (dst_patch_shape[0] * 2 - 1) * (dst_patch_shape[1] * 2 - 1)
+                    src_size = int((src_num_pos - num_extra_tokens) ** 0.5)
```
Comment (MeowZheng) [inline mmseg/models/backbones/beit.py:440]: `# src_size is the distance of relative position embedding: 2*w_src-1`

File: mmseg/models/backbones/beit.py [hunk 12]
```diff
@@ -0,0 +1,563 @@
...
+                    dst_size = int((dst_num_pos - num_extra_tokens) ** 0.5)
```
Comment (MeowZheng) [inline mmseg/models/backbones/beit.py:441]: `# dst_size is the distance of relative position embedding: 2*w_dst-1`

**Pull request closed without merging**